csharp/src/Google.Protobuf/CodedOutputWriter.cs

```diff
@@ -0,0 +1,662 @@
+            {
+                EnsureMore(4);
+            }
+
+            span[buffered++] = b1;
+            span[buffered++] = b2;
+            span[buffered++] = b3;
+            span[buffered++] = b4;
+        }
+
+        /// <summary>
+        /// Writes the given five-byte tag directly to the stream.
+        /// </summary>
+        /// <param name="b1">The first byte of the encoded tag</param>
+        /// <param name="b2">The second byte of the encoded tag</param>
+        /// <param name="b3">The third byte of the encoded tag</param>
+        /// <param name="b4">The fourth byte of the encoded tag</param>
+        /// <param name="b5">The fifth byte of the encoded tag</param>
+        public void WriteRawTag(byte b1, byte b2, byte b3, byte b4, byte b5)
+        {
+            if (buffered + 4 > span.Length)
+            {
+                EnsureMore(5);
+            }
+
+            span[buffered++] = b1;
+            span[buffered++] = b2;
+            span[buffered++] = b3;
+            span[buffered++] = b4;
+            span[buffered++] = b5;
+        }
+        #endregion
+
+        #region Underlying writing primitives
+        /// <summary>
+        /// Writes a 32 bit value as a varint. The fast route is taken when
+        /// there's enough buffer space left to whizz through without checking
+        /// for each byte; otherwise, we resort to checking size each byte.
+        /// </summary>
+        internal void WriteRawVarint32(uint value)
+        {
+            // Optimize for the common case of a single byte value
+            if (value < 128 && buffered < span.Length)
+            {
+                span[buffered++] = (byte)value;
+            }
+            else
+            {
+                while (value > 127)
+                {
+                    Ensure(1);
+                    span[buffered++] = (byte)((value & 0x7F) | 0x80);
+                    value >>= 7;
+                }
+
+                Ensure(1);
+                span[buffered++] = (byte)value;
+            }
+        }
+
+        internal void WriteRawVarint64(ulong value)
+        {
+            // Optimize for the common case of a single byte value
+            if (value < 128 && buffered < span.Length)
+            {
+                span[buffered++] = (byte)value;
+            }
+            else
+            {
+                while (value > 127)
+                {
+                    Ensure(1);
+                    span[buffered++] = (byte)((value & 0x7F) | 0x80);
+                    value >>= 7;
+                }
+
+                Ensure(1);
+                span[buffered++] = (byte)value;
+            }
+        }
+
+        internal void WriteRawLittleEndian32(uint value)
+        {
+            Ensure(4);
+
+            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(buffered), value);
+
+            buffered += 4;
+        }
+
+        internal void WriteRawLittleEndian64(ulong value)
+        {
+            Ensure(8);
```

Comment:
nit: I've seen const int, sizeof and actual numbers. Maybe declare these constants in a consistent way?